### ISSUE TITLE
Fatal Error: Listener socket is closed

### ISSUE THREAD

**Issue by paralin** [statianzo/Fleck#97]
Hi all,

When a lot of connections try to connect right as the server starts up, I get the following error: 

```
6/4/2014 6:40:00 PM [Error] Listener socket is closed System.AggregateException:  ---> System.Net.Sockets.SocketException: The socket is not connected  
  at System.Net.Sockets.Socket.get_RemoteEndPoint () [0x00000] in <filename unknown>:0   
  at Fleck.SocketWrapper.get_RemoteIpAddress () [0x00000] in <filename unknown>:0   
  at Fleck.WebSocketServer.OnClientConnect (ISocket clientSocket) [0x00000] in <filename unknown>:0   
  at Fleck.SocketWrapper+<Accept>c__AnonStorey2.<>m__1 (System.Threading.Tasks.Task`1 t) [0x00000] in <filename unknown>:0   
  at System.Threading.Tasks.TaskActionInvoker+ActionTaskInvoke`1[Fleck.ISocket].Invoke (System.Threading.Tasks.Task owner, System.Object state, System.Threading.Tasks.Task context) [0x00000] in <filename unknown>:0   
  at System.Threading.Tasks.Task.InnerInvoke () [0x00000] in <filename unknown>:0   
  at System.Threading.Tasks.Task.ThreadStart () [0x00000] in <filename unknown>:0   
  --- End of inner exception stack trace ---  
 --> (Inner exception 0) System.Net.Sockets.SocketException: The socket is not connected  
  at System.Net.Sockets.Socket.get_RemoteEndPoint () [0x00000] in <filename unknown>:0   
  at Fleck.SocketWrapper.get_RemoteIpAddress () [0x00000] in <filename unknown>:0   
  at Fleck.WebSocketServer.OnClientConnect (ISocket clientSocket) [0x00000] in <filename unknown>:0   
  at Fleck.SocketWrapper+<Accept>c__AnonStorey2.<>m__1 (System.Threading.Tasks.Task`1 t) [0x00000] in <filename unknown>:0   
  at System.Threading.Tasks.TaskActionInvoker+ActionTaskInvoke`1[Fleck.ISocket].Invoke (System.Threading.Tasks.Task owner, System.Object state, System.Threading.Tasks.Task context) [0x00000] in <filename unknown>:0   
  at System.Threading.Tasks.Task.InnerInvoke () [0x00000] in <filename unknown>:0   
  at System.Threading.Tasks.Task.ThreadStart () [0x00000] in <filename unknown>:0 
```

**Comment by bennidhamma** [comment]
I think we are experiencing this as well. During busy times, our socket server is after about 2,000 connection attempts. But at night, when we are less busy, we are able to run for considerably longer.

**Comment by jespertheend** [comment]
From [msdn](https://msdn.microsoft.com/en-us/library/system.net.sockets.socket.remoteendpoint(v=vs.110).aspx)

> The RemoteEndPoint is set after a call to either Accept or Connect. If you try to access this property earlier, RemoteEndPoint will throw a SocketException. If you receive a SocketException, use the SocketException.ErrorCode property to obtain the specific error code. After you have obtained this code, refer to the Windows Sockets version 2 API error code documentation in the MSDN library for a detailed description of the error.

SocketWrapper.cs contains the following code

        public string RemoteIpAddress
        {
            get
            {
                var endpoint = _socket.RemoteEndPoint as IPEndPoint;
                return endpoint != null ? endpoint.Address.ToString() : null;
            }
        }

I assume this is called without being certain if Accept or Connect has been called. I'm not sure though, I'll see if I can write a fix for this.

**Comment by jespertheend** [comment]
I tried to check if `_socket.Connected` and returning `null` and `-1` when it returned false, but this was still causing errors. I also tried catching the `SocketException` but this caused the listener to not restart, so new connections made to the socket would fail. I guess there's no need to fix this error since `server.RestartAfterListenError = true;` already fixes this, but it would be nice to make the logs a bit cleaner.
